x-pack/plugins/security/server/authentication/providers/saml.ts

```diff
@@ -307,11 +347,17 @@ export class SAMLAuthenticationProvider extends BaseAuthenticationProvider {
       this.logger.debug('Login has been performed with SAML response.');
       return AuthenticationResult.redirectTo(
         stateRedirectURL || `${this.options.basePath.get(request)}/`,
-        { state: { username, accessToken, refreshToken } }
+        { state: { username, accessToken, refreshToken, realm: this.realm } }
       );
     } catch (err) {
       this.logger.debug(`Failed to log in with SAML response: ${err.message}`);
-      return AuthenticationResult.failed(err);
+
+      // Since we don't know upfront what realm is targeted by the Identity Provider initiated login
+      // there is a chance that it failed because of realm mismatch and hence we should return
+      // `notHandled` and give other SAML providers a chance to properly handle it instead.
+      return isIdPInitiatedLogin
```

Comment:
__note:__ more correct solution here would be to not send `realm` to Elasticsearch and:

* If request failed - return `failed()`
* If request succeeded and returned realm equal to the one this provider relies on - return `succeeded()`
* If request succeeded, but returned realm __not__ equal to the one this provider relies on - invalidate created tokens and return `notHandled`

But I really don't like the `invalidate created tokens` step.

The other solution I was thinking about was to re-use the same provider instance for all enabled providers of the same type. This solution wouldn't have the problem I'm trying to solve here, but it has the set of its own issues. Happy to discuss if you think current solution isn't acceptable.

Would do you think?

Reply:
Does having to support `xpack.security.public.*` settings in 7.x affect this at all? Perhaps we could discuss this specific flow over Zoom. I'm not sure I'm following all of the complexities entirely.

Reply:
> Does having to support xpack.security.public.* settings in 7.x affect this at all?

As I replied to another comment I expect to not have problem with `xpack.security.public.*` if we forbid multiple SAML providers if there any without `realm` specified (basically when Kibana is already configured to use SAML with old config format without specified `realm`).

> Perhaps we could discuss this specific flow over Zoom. I'm not sure I'm following all of the complexities entirely.

Sure, let's discuss over Zoom.

Reply:
What we discussed on Zoom makes sense, no concerns on this point.